packages/@lwc/template-compiler/src/parser/parser.ts

```diff
@@ -134,33 +167,138 @@ export default class ParserCtx {
-        const currentScope = this.currentScope() || [];
-        const sibling = currentScope.find(predicate);
-        return sibling || null;
+    findInCurrentElementScope<A extends ParentNode>(
+        predicate: (node: ParentNode) => node is A
+    ): A | null {
+        const currentScope = this.currentElementScope() || [];
+        return currentScope.find(predicate) || null;
     }
 
-    beginScope(): void {
-        this.scopes.push([]);
+    beginElementScope(): void {
+        this.elementScopes.push([]);
     }
 
-    endScope(): void {
-        this.scopes.pop();
+    endElementScope(): ParentNode | undefined {
+        const scope = this.elementScopes.pop();
+        return scope ? scope[0] : undefined;
     }
 
-    addNodeCurrentScope(node: ParentNode): void {
-        const currentScope = this.currentScope();
+    addNodeCurrentElementScope(node: ParentNode): void {
+        const currentScope = this.currentElementScope();
 
         /* istanbul ignore if */
         if (!currentScope) {
-            throw new Error("Can't invoke addNodeCurrentScope if there is no current scope");
+            throw new Error("Can't invoke addNodeCurrentElementScope if there is no current scope");
         }
 
         currentScope.push(node);
     }
 
-    private currentScope(): ParentNode[] | undefined {
-        return this.scopes[this.scopes.length - 1];
+    hasSeenSlot(name: string): boolean {
+        return this.seenSlotsFromAncestorIfTree().has(name);
+    }
+
+    addSeenSlot(name: string): void {
+        const currentSeenSlots = this.seenSlotsFromAncestorIfTree();
+        if (currentSeenSlots) {
+            currentSeenSlots.add(name);
+        } else {
+            this.seenSlots.add(name);
+        }
+    }
+
+    private currentElementScope(): ParentNode[] | undefined {
+        return this.elementScopes[this.elementScopes.length - 1];
+    }
+
+    beginSiblingScope() {
+        this.siblingScopes.push({
+            ancestorIfContext: this.currentIfContext() || this.ancestorIfContext(),
+        });
+    }
+
+    endSiblingScope() {
+        this.siblingScopes.pop();
+    }
+
+    beginIfChain(node: IfBlock) {
+        const currentSiblingContext = this.currentSiblingContext();
+        if (!currentSiblingContext) {
+            throw new Error('Cannot invoke beginIfChain if there is currently no sibling context');
+        }
+
+        const currentIfContext = this.currentIfContext();
+        if (currentIfContext) {
+            throw new Error(
+                'Should not invoke beginIfChain if an if context already exists. First end the current chain before starting a new one.'
+            );
+        }
+
+        const previouslySeenSlots = this.seenSlotsFromAncestorIfTree();
+        currentSiblingContext.ifContext = {
+            currentNode: node,
+            seenSlots: [new Set<string>(previouslySeenSlots)],
+        };
+    }
+
+    updateIfChain(node: ElseifBlock | ElseBlock) {
```

Comment:
Suggestion: what if we rename this to something like `appendElseOrElseif`? `update` seems kind of generic.